core/store/hashtree/mem/tree.go

```diff
@@ -0,0 +1,477 @@
+	if path != nil {
+		path.leaf = n
+	}
+
+	return nil
+}
+
+// Insert implements mem.TreeNode. It replaces the empty node by a leaf node
+// that contains the key and the value.
+func (n *EmptyNode) Insert(key, value []byte, bi *big.Int) TreeNode {
+	return NewLeafNode(n.depth, key, value)
+}
+
+// Delete implements mem.TreeNode. It ignores the delete as an empty node
+// already means the key is missing.
+func (n *EmptyNode) Delete(key *big.Int) TreeNode {
+	return n
+}
+
+// Prepare implements mem.TreeNode. It updates the hash of the node and return
+// the digest.
+func (n *EmptyNode) Prepare(nonce []byte, prefix *big.Int, fac crypto.HashFactory) ([]byte, error) {
+	h := fac.New()
+
+	data := make([]byte, 1+len(nonce)+prefix.BitLen()+DepthLength)
+	cursor := 1
+	data[0] = emptyNodeType
+	copy(data[cursor:], nonce)
+	cursor += len(nonce)
+	copy(data[cursor:], prefix.Bytes())
+	cursor += prefix.BitLen()
+	copy(data[cursor:], int2buffer(n.depth))
+
+	_, err := h.Write(data)
+	if err != nil {
+		return nil, xerrors.Errorf("empty node failed: %v", err)
+	}
+
+	n.hash = h.Sum(nil)
+
+	return n.GetHash(), nil
+}
+
+// Visit implements mem.TreeNode. It executes the callback with the node.
+func (n *EmptyNode) Visit(fn func(TreeNode)) {
+	fn(n)
+}
+
+// Clone implements mem.TreeNode. It returns a deep copy of the empty node.
+func (n *EmptyNode) Clone() TreeNode {
+	return NewEmptyNode(n.depth)
+}
+
+// InteriorNode is a node with two children.
+type InteriorNode struct {
+	hash  []byte
+	depth uint16
+	left  TreeNode
+	right TreeNode
+}
+
+// NewInteriorNode creates a new interior node with two empty nodes as children.
+func NewInteriorNode(depth uint16) *InteriorNode {
+	return &InteriorNode{
+		depth: depth,
+		left:  NewEmptyNode(depth + 1),
+		right: NewEmptyNode(depth + 1),
+	}
+}
+
+// GetHash implements mem.TreeNode. It returns the hash of the node.
+func (n *InteriorNode) GetHash() []byte {
+	return append([]byte{}, n.hash...)
+}
+
+// GetType implements mem.TreeNode. It returns the interior node type.
+func (n *InteriorNode) GetType() byte {
+	return interiorNodeType
+}
+
+// Search implements mem.TreeNode. It recursively search for the value in the
+// correct child.
+func (n *InteriorNode) Search(key *big.Int, path *Path) []byte {
+	if key.Bit(int(n.depth)) == 0 {
+		if path != nil {
+			path.interiors = append(path.interiors, n.right.GetHash())
+		}
+
+		return n.left.Search(key, path)
+	}
+
+	if path != nil {
+		path.interiors = append(path.interiors, n.left.GetHash())
+	}
+
+	return n.right.Search(key, path)
+}
+
+// Insert implements mem.TreeNode. It inserts the key/value pair to the right
+// path.
+func (n *InteriorNode) Insert(key, value []byte, bi *big.Int) TreeNode {
+	if bi.Bit(int(n.depth)) == 0 {
+		n.left = n.left.Insert(key, value, bi)
+	} else {
+		n.right = n.right.Insert(key, value, bi)
+	}
+
+	return n
+}
+
+// Delete implements mem.TreeNode. It deletes the key from the right path.
```

Comment:
```suggestion
// Delete implements mem.TreeNode. It deletes the key from the right of left path, based on the bi path.
```

The naming is fragile here, in this function we have a `key` which is a `*big.Int`, but in the previous `Insert()` one we have a `key` that is a `[]byte` and then another `*big.Int` argument which has the same utility as here but is called there `bi`.

Reply:
This is a global remark: it can be hard to get around all the different notions of `key` `index` `prefix` `path` `value` `depth` that sometimes use `[]byte` or `*big.Int`, or is even converted to `int`.

I would find really useful to have a small example of the tree with what and how things are stored. Sometimes just having a small representation does 90% of the job (a bit like the routing tree: https://github.com/dedis/dela/blob/master/mino/minogrpc/routing/routing.go#L282)

Reply:
In all cases, it's still the key but it can actually be always a *big.Int for simplicity. For the example, the implementation is based on a paper that explains everything so I'm not really keen to write a different (and probably bad) explanation on how it works where the paper is really smooth in that regard.